lib2/client/okta.go

```diff
@@ -0,0 +1,667 @@
+// Client for making requests to Okta APIs
+package client
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/cookiejar"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"golang.org/x/net/publicsuffix"
+
+	"github.com/99designs/keyring"
+	"github.com/segmentio/aws-okta/lib"
+	"github.com/segmentio/aws-okta/lib/mfa"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	OktaServerUs      = "okta.com"
+	OktaServerEmea    = "okta-emea.com"
+	OktaServerPreview = "oktapreview.com"
+	OktaServerDefault = OktaServerUs
+
+	// deprecated; use OktaServerUs
+	OktaServer = OktaServerUs
+
+	Timeout = time.Duration(60 * time.Second)
+)
+
+type OktaUser struct {
+	Username string `json:"username"`
+	Password string `json:"password"`
+}
+type OktaClient struct {
+	creds        OktaCredential
+	UserAuth     *OktaUserAuthn
+	DuoClient    *lib.DuoClient
+	SessionToken string
+	Expiration   time.Time
+	CookieJar    http.CookieJar
+	BaseURL      *url.URL
+	MFAConfig    MFAConfig
+	Keyring      *keyring.Keyring
+	client       http.Client
+}
+
+type MFAConfig struct {
+	Provider   string // Which MFA provider to use when presented with an MFA challenge
+	FactorType string // Which of the factor types of the MFA provider to use
+	DuoDevice  string // Which DUO device to use for DUO MFA
+}
+
+// type: OktaCredential struct stores Okta credentials and domain information that will
+// be used by OktaClient when making API calls to Okta
+type OktaCredential struct {
+	Username string
+	Password string
+	Domain   string
+}
+
+// Checks the validity of OktaCredential and should be called before
+// using the credentials to make API calls.
+//
+// This public method will only validate that credentials exist, it will NOT
+// validate them for correctness. To validate correctness an OktaClient must be
+// used to make a request to Okta.
+func (c *OktaCredential) IsValid() bool {
+	return c.Username != "" && c.Password != "" && c.Domain != ""
+}
+
+// Will fetch your Okta username, password, and domain from your keyring secret
+// backend.
+//
+// Will get the default credentials stored under the `okta-creds` key.
+//
+// feat-request: add support for getting additional sets of credentials.
+// The interface for this functionality needs to be defined, it's
+// possible to then implement alternative credential backends
+// to flexibly support alternative implementations.
+func GetOktaCredentialFromKeyring(kr keyring.Keyring) (OktaCredential, error) {
```

Comment:
Can we factor keyring out of the client code and into a simple interface? In the same way I did `sessioncache` I suppose.

Just thinking if you're using `client` from server code, you probably aren't going to want to use `keyring`.

Reply:
This helper function is meant as a compromise for getting the okta crendentials from the keyring. Maybe it should be moved to `cmd`?

for the session caching I would prefer something like `sessioncache`, maybe if there is a session cache that returns `[]byte` instead of a struct?

something like:
```
type SessionCacheInterface interface {
  Get(key string) ([]byte, error)
  Put(key string, data []byte, name string) (error)
}
```

Reply:
Yep, I wasn't necessarily proposing copying `SessionCacheInterface` verbatim, just copying the pattern.

Moving keyring functionality to `cmd` seems sensible. It could be a subpackage or not, depending on how big it is.

Reply:
I'm implemented the session cache for OktaClient. 

Keeping the helper function where it is seems fine for now, I've reconsidered and it doesn't seem like we need to move it.